### PR TITLE
Implement the Mehrstellenverfahren exercise

### PR DESCRIPTION
**Pull request by TobiasRohner**

### CODE REVIEW

File: developers/Mehrstellenverfahren/mastersolution/mehrstellenverfahren.cc [hunk 1]
```diff
@@ -1,11 +1,133 @@
...
+  for (unsigned int M : Ms) {
+    errs.push_back(compgriderr(M));
+  }
+#else
+  //====================
+  // Your code goes here
+  //====================
+#endif
+  // Print the collected data out in a table
+  printf("| M   | h       | error        |\n");
```
Comment (hiptmair) [inline developers/Mehrstellenverfahren/mastersolution/mehrstellenverfahren.cc:123]: `printf` in C++? Why not cout with iomanip, which would be the better style?

**Review by hiptmair — CHANGES_REQUESTED** [review]
The initialization of the matrix does not seem to be efficient, see my comment in the code.

File: developers/Mehrstellenverfahren/mastersolution/mehrstellenverfahren.cc (hunk 1 above)
Comment (TobiasRohner) [reply developers/Mehrstellenverfahren/mastersolution/mehrstellenverfahren.cc:123]: I thought about using iomanip when implementing this function but settled for printf due to readability of the code. Printing a single row of the table with iomanip would be done like this: `std::cout << "| " << std::setw(3) << M << " | " << std::fixed << std::setprecision(5) << h << " | " << std::scientific << err << " |\n"`. In my opinion, this is quite unreadable and on has to stare at it for quite a hile to understand what is happening. I can change it if you want to, but would suggest to switch to `std::format` as soon as we upgrade to C++20.

File: developers/Mehrstellenverfahren/mastersolution/mehrstellenverfahren.cc [hunk 2]
```diff
@@ -1,11 +1,133 @@
...
+#include <vector>
+
+namespace mehrstellenverfahren {
+
+Eigen::SparseMatrix<double> compMehrstellenA(unsigned int M) {
+  Eigen::SparseMatrix<double> A(M * M, M * M);
+#if SOLUTION
+  // We have at most 9nnz per row of A and A has M^2 rows
+  const int nnz = 9 * M * M;
+  A.reserve(nnz);
```
[the remark this answers is not in the document]
Comment (TobiasRohner) [reply developers/Mehrstellenverfahren/mastersolution/mehrstellenverfahren.cc:22]: According to the Eigen documentation, the `reserve` function either takes a scalar value describing the total number of non-zero entries or a vector containing the number of non-zero entries in the i-th column/row in its i-th component. So the amount of memory that is allocated is correct. The reason for the inefficiency was that the speed of insertions is strongly dependent on their order. For a column-major matrix (the default in Eigen), entries in a column have to be inserted in order. This is now fixed and the code runs significantly faster.

File: developers/Mehrstellenverfahren/mastersolution/mehrstellenverfahren.cc (hunk 1 above)
Comment (hiptmair) [reply developers/Mehrstellenverfahren/mastersolution/mehrstellenverfahren.cc]: OK, you have a point :smile:

**Pull request merged**